platform/Callback.h

```diff
@@ -35,209 +42,441 @@ namespace mbed {
      *  @param arg      Pointer argument to function
      */
-    template<typename T, typename U>
-    Callback(R(*func)(T *, ArgTs...), U *arg)
+    template<typename Fn, typename BoundArg, typename std::enable_if_t<mstd::is_invocable_r<R, Fn, BoundArg, ArgTs...>::value, int> = 0>
+    Callback(Fn func, BoundArg arg) : CallbackBase()
     {
-        generate(function_context<R(*)(T *, ArgTs...), T>(func, arg));
+        generate([func, arg](ArgTs... args) {
+            return detail::invoke_r<R>(func, arg, std::forward<ArgTs>(args)...);
+        });
     }
 
-    /** Create a Callback with a static function and bound pointer
-     *  @param func     Static function to attach
-     *  @param arg      Pointer argument to function
+    // *INDENT-OFF*
+    /** Create a Callback with a function object
+     *  @param f Function object to attach
+     *  @note The function object is limited to a a few words of storage
      */
-    template<typename T, typename U>
-    Callback(R(*func)(const T *, ArgTs...), const U *arg)
+    template <typename F,
+            typename std::enable_if_t<
+                    !detail::can_null_check<F>::value &&
+                    mstd::is_invocable_r<R, F, ArgTs...>::value, int> = 0>
+    Callback(F f) : CallbackBase()
     {
-        generate(function_context<R(*)(const T *, ArgTs...), const T>(func, arg));
+        static_assert(std::is_copy_constructible<F>::value, "Callback F must be CopyConstructible");
+        generate(std::move(f));
     }
 
-    /** Create a Callback with a static function and bound pointer
-     *  @param func     Static function to attach
-     *  @param arg      Pointer argument to function
+    /** Create a Callback with a function pointer
+     *  @param f Function pointer to attach
      */
-    template<typename T, typename U>
-    Callback(R(*func)(volatile T *, ArgTs...), volatile U *arg)
+    template <typename F,
+            typename std::enable_if_t<
+                    detail::can_null_check<F>::value &&
+                    mstd::is_invocable_r<R, F, ArgTs...>::value, int> = 0>
+    Callback(F f) : CallbackBase()
     {
-        generate(function_context<R(*)(volatile T *, ArgTs...), volatile T>(func, arg));
+        static_assert(std::is_copy_constructible<F>::value, "Callback F must be CopyConstructible");
+        if (!f) {
+            clear();
+        } else {
+            generate(std::move(f));
+        }
     }
+    // *INDENT-ON*
 
-    /** Create a Callback with a static function and bound pointer
-     *  @param func     Static function to attach
-     *  @param arg      Pointer argument to function
+#if MBED_CONF_PLATFORM_CALLBACK_NONTRIVIAL
+    /** Destroy a callback
      */
-    template<typename T, typename U>
-    Callback(R(*func)(const volatile T *, ArgTs...), const volatile U *arg)
+    ~Callback()
     {
-        generate(function_context<R(*)(const volatile T *, ArgTs...), const volatile T>(func, arg));
+        destroy();
     }
+#endif
```

Comment:
Would it be possible to have destructor mark as default if callback is trivial ?

Reply:
Yes, guess it's worthwhile for documentation purposes. I think we're having `= default` on the copy constructor unnecessarily already.

Reply:
Yes, we use default for all the other special member functions.